### PR TITLE
Add support for incremental ThinLTO + ThinLTO with mold

### PR DESCRIPTION
**Pull request by ishitatsuyuki**
Closes: #10311

### CODE REVIEW

File: mesonbuild/compilers/mixins/clang.py [hunk 1]
```diff
@@ -148,17 +149,25 @@ def get_coverage_link_args(self) -> T.List[str]:
...
-            # Thin LTO requires the use of gold, lld, ld64, or lld-link
-            if not isinstance(self.linker, (AppleDynamicLinker, ClangClDynamicLinker, LLVMDynamicLinker, GnuGoldDynamicLinker)):
-                raise mesonlib.MesonException(f"LLVM's thinLTO only works with gnu gold, lld, lld-link, and ld64, not {self.linker.id}")
+            # ThinLTO requires the use of gold, lld, ld64, lld-link or mold 1.1+
+            if isinstance(self.linker, (MoldDynamicLinker)):
+                # https://github.com/rui314/mold/commit/46995bcfc3e3113133620bf16445c5f13cd76a18
+                if not mesonlib.version_compare(self.linker.version, '>=1.1'):
+                    raise mesonlib.MesonException(f"LLVM's ThinLTO requires mold 1.1+")
+            elif not isinstance(self.linker, (AppleDynamicLinker, ClangClDynamicLinker, LLVMDynamicLinker, GnuGoldDynamicLinker)):
+                raise mesonlib.MesonException(f"LLVM's ThinLTO only works with gnu gold, lld, lld-link, and ld64, not {self.linker.id}")
```
Comment (tristan957) [inline mesonbuild/compilers/mixins/clang.py:158]: Maybe remove the mention of GNU here? Or capitalize it.

**Comment by tristan957** [comment]
Sucks that a lot of functionality exists only in the ninja backend due to the way the backends are abstracted.

Thanks for submitting the PR! You requested the feature and you put in the work. Awesome stuff.

File: mesonbuild/compilers/mixins/clang.py (hunk 1 above)
Comment (ishitatsuyuki) [reply mesonbuild/compilers/mixins/clang.py]: Message updated.

**Comment by ishitatsuyuki** [comment]
Thanks for the kind words! I've addressed your comment.

**Comment by jpakkane** [comment]
Instead of adding a new feature, could we make it always automatically work?

**Comment by ishitatsuyuki** [comment]
As mentioned in the issue, I made this an opt-in because it’s pure overhead for one-time builds (e.g. packaging).

**Comment by eli-schwartz** [comment]
This should probably get release notes, particularly for adding a cache.

Hmm, ccache does work even after a --wipe because it inherently uses a per-user cache, not a per-build cache. Would it make sense to accept a path for thinlto in addition to true/false?

**Comment by ishitatsuyuki** [comment]
> This should probably get release notes, particularly for adding a cache.

Will do.

> Hmm, ccache does work even after a --wipe because it inherently uses a per-user cache, not a per-build cache. Would it make sense to accept a path for thinlto in addition to true/false?

Sure, I'll get another option in to specify that.

File: docs/markdown/Builtin-options.md [hunk 2]
```diff
@@ -123,23 +123,24 @@ no options.
...
+| b_coverage      | false          | true, false                                                   | Enable coverage tracking                                                       |
+| b_lundef        | true           | true, false                                                   | Don't allow undefined symbols when linking                                     |
+| b_lto           | false          | true, false                                                   | Use link time optimization                                                     |
+| b_lto_threads   | 0              | Any integer*                                                  | Use multiple threads for lto. *(Added in 0.57.0)*                              |
+| b_lto_mode      | default        | default, thin                                                 | Select between lto modes, thin and default. *(Added in 0.57.0)*                |
+| b_thinlto_cache | default        | true, false                                                   | Enable LLVM's ThinLTO cache for faster incremental builds. *(Added in 0.63.0)* |
```
Comment (Tachi107) [inline docs/markdown/Builtin-options.md:136]: Here the default value is "default". Did you mean to write "false"?

**Comment by ishitatsuyuki** [comment]
Sorry for the delay, I've now added an option to specify the cache directory. (Plus a version bump for "added in" docs to compensate for the delay :stuck_out_tongue:)

One thing I doesn't understand is that the `-Wl,--thinlto-cache-dir=<dir>` option is only updated by `--wipe`, not `--reconfigure`. Any idea what I can do to fix that?

**Comment by ishitatsuyuki** [comment]
It has been a week since I last updated the patch. Could we make progress on getting a review consensus? Any idea about the `--reconfigure` issue I mentioned?

**Comment by eli-schwartz** [comment]
I don't know what the reconfigure issue is and I can't investigate right now, but have you tried printf debugging along the call route you added and checking what changes during a reconfigure?

**Comment by tristan957** [comment]
@dcbaker maybe you know something about this?

**Comment by ishitatsuyuki** [comment]
Do we have any concerns apart from the general reconfigure stuff (which seems to be an old thing)? I'd like to get this merged, so that I don't need to source my venv everytime I want to reconfigure my local Mesa build.

**Comment by eli-schwartz** [comment]
> This should probably get release notes, particularly for adding a cache.

**Comment by ishitatsuyuki** [comment]
I added an changelog entry.

**Pull request merged**